src/lib/components/FornaContainer.react.js

```diff
@@ -0,0 +1,274 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import {FornaContainer as PreFornaContainer} from 'fornac';
+import * as R from 'ramda';
+
+/**
+ * This is a FornaContainer component.
+ */
+export default class FornaContainer extends Component {
+    constructor(props) {
+        super(props);
+        this.renderNewSequences = this.renderNewSequences.bind(this);
+    }
+
+    componentDidMount() {
+        const {id, height, width} = this.props;
+
+        this._fornaContainer = new PreFornaContainer('#' + id, {
+            initialSize: [width, height],
+        });
+        this.renderNewSequences();
+    }
+
+    renderNewSequences() {
+        const {sequences} = this.props;
+
+        if (this._fornaContainer) {
+            this._fornaContainer.clearNodes();
+
+            sequences.forEach(seq => {
+                const unpackedOptions = Object.assign({}, seq.options, {
+                    sequence: seq.sequence,
+                    structure: seq.structure,
+                });
+                this._fornaContainer.addRNA(seq.structure, unpackedOptions);
+            });
+        }
+    }
+
+    shouldComponentUpdate(nextProps) {
+        const {sequences, colorScheme, nodeFillColor} = this.props;
+
+        if (Boolean(sequences) !== Boolean(nextProps.sequences)) {
+            return true;
+        }
+
+        if (sequences.length !== nextProps.sequences.length) {
+            return true;
+        }
+
+        const sequence_compare = R.zip(sequences, nextProps.sequences);
+
+        if (
+            sequence_compare.some(
+                seq =>
+                    seq[0].sequence !== seq[1].sequence ||
+                    seq[0].structure !== seq[1].structure
+            )
+        ) {
+            return true;
+        }
+
+        if (
+            sequence_compare.some(
+                seq =>
+                    Boolean(seq[0].options) !== Boolean(seq[1].options) ||
+                    (Boolean(seq[0].options) &&
+                        Boolean(seq[1].options) &&
+                        R.symmetricDifference(
+                            Object.keys(seq[0].options),
+                            Object.keys(seq[1].options)
+                        ).length > 0)
+            )
+        ) {
+            return true;
+        }
+
+        if (
+            sequence_compare.some(
+                seq =>
+                    Boolean(seq[0].options) &&
+                    Boolean(seq[1].options) &&
+                    Object.keys(seq[0].options).some(
+                        _option =>
+                            seq[0].options._option !== seq[1].options._option
+                    )
+            )
+        ) {
+            return true;
+        }
```

Comment:
seems like this could be more efficient combining all the loops into one. Also doesn't the whole `symmetricDifference` then looping through options - in fact even the `Boolean(...) !== Boolean(...)` amount to "are the two objects equal"? Which is what `R.equals` is made for:

```js
sequence_compare.some(
    ([prev, next]) =>
        prev.sequence !== next.sequence ||
        prev.structure !== next.structure ||
        !R.equals(prev.options, next.options)
)
```

or in fact that's everything in the `sequences` prop, right? You just want to know if there's any change anywhere in that prop... So can't this WHOLE thing - all the way up to `if (Boolean(sequences) !== Boolean(nextProps.sequences))` - be replaced by
```js
if (!R.equals(sequences, nextProps.sequences)) {
    return true;
}
```
? Apologies for not noticing that earlier, before you went through a bunch of iterations here...

Reply:
Yes, that makes sense! I didn't know about `R.equals`, looks like I can use it in a lot of other components to (like `SequenceViewer`).